src/api/use-subscription.ts

```diff
@@ -25,40 +28,66 @@ export function useSubscription({
   const [state, setState] = useState<SubscriptionState>('NOT_STARTED');
   const [webSocket, setWebSocket] = useState<WebSocket | null>(null);
 
+  const [reconnectCount, setReconnectCount] = useState(0);
+  const reconnect = () => setReconnectCount(reconnectCount + 1);
+
+  const [retryCount, setRetryCount] = useState(0);
+  const retryWithExponentialBackoff = () => {
+    const delay = Math.pow(2, retryCount) * 1000;
+    return setTimeout(() => setRetryCount(retryCount + 1), delay);
+  };
+
   useEffect(() => {
+    let retryTimeout: NodeJS.Timeout | null = null;
     if (url) {
       const webSocket = new WebSocket(url);
+
       webSocket.onmessage = (event) => {
         setState(getSubscriptionStateCode(webSocket.readyState));
         onMessage && onMessage(event);
       };
+
       webSocket.onerror = (event) => {
         setState(getSubscriptionStateCode(webSocket.readyState));
         Bugsnag.notify(`WebSocket error "${event.message}" with url: ${url}`);
         onError && onError(event);
       };
+
       webSocket.onclose = (event) => {
         setState(getSubscriptionStateCode(webSocket.readyState));
         Bugsnag.leaveBreadcrumb(`WebSocket closed with url: ${url}`);
+
+        // Reconnect immediately if close event is end of stream, otherwise use
+        // exponetial backoff to retry.
+        if (event.code === 1001 && reconnectCount < MAX_NUMBER_OF_RECONNECTS) {
+          reconnect();
+        } else if (retryCount < MAX_NUMBER_OF_RETRIES) {
+          retryTimeout = retryWithExponentialBackoff();
+        }
+
         onClose && onClose(event);
       };
+
       webSocket.onopen = () => {
         setState(getSubscriptionStateCode(webSocket.readyState));
         Bugsnag.leaveBreadcrumb(`WebSocket opened with url: ${url}`);
+        setRetryCount(0);
         onOpen && onOpen();
       };
+
       setWebSocket(webSocket);
     }
 
     // Cleanup
     return () => {
+      if (retryTimeout) clearTimeout(retryTimeout);
       if (webSocket) {
         webSocket.onclose = null;
         onClose && onClose({code: 1000, reason: 'Cleanup'});
         webSocket.close();
       }
     };
-  }, [url]);
+  }, [url, reconnectCount, retryCount]);
```

Comment:
Can't have an `useEffect` which has `reconnectCount` and `retryCount` in the dependency list, while using `setReconnectCount` and `setRetryCount` in the body. Need to refactor somehow. Maybe the counts can be refs instead of state if changing them isn't ment to cause a rerender.

Reply:
I realize there's a bit of implicit behavior here, that might be hard to spot when reading the code. 

`reconnectCount` and `retryCount` have two purposes:

1. They are used to keep track of how many times the app should try to retry/reconnect
2. They are used to trigger the retry/reconnnect itself, because updating the value also triggers the useEffect. Changing them to refs would defeat this purpose.

This seems to work very well in practice, but am I solving it in a completely roundabout-way? It's possible that I'm missing another good way to solve this, or should I update the name of the variables?

Reply:
You can't have a state value in the dependency list of a `useEffect` which modifies that state value. Basic example

```ts
const [count, setCount] = useState(0);
// Infinite loop
useEffect(() => {
  setCount(count + 1)
}, [count]);
```

Even though some logic in this `useEffect` hinders the infinite lopp, the rule that setting state for a state variable that is in the dependency list is a clear no-no, and should be avoided.

Also note that having the `reconnectCount` and `retryCount` in the dependency list makes the cleanup function run every time one of those changes. So when everything works and `setRetryCount(0)` is run, then the cleanup triggers closing the connection and reconnecting again. (Not tested, just what I think will happen when reading the code)

Reply:
You are right, I didn't find a well established source that setting state that is in the dependency list is strictly forbidden. As long as measures are done so it isn't causing a loop or unnecessary renders. I will soften my stance on it, but still be "on guard" when I see it happening since it can be the cause for some bugs :)

And good refactor! :+1: